File: 188.c
#include <stdio.h>
#include <stdint.h>

uint64_t mul_mod(uint64_t a, uint64_t b, uint64_t m)
{
   uint64_t d = 0, mp2 = m >> 1;
   int i;
   if (a >= m) a %= m;
   if (b >= m) b %= m;
   for (i = 0; i < 64; ++i)
   {
       d = (d > mp2) ? (d << 1) - m : d << 1;
       if (a & 0x8000000000000000ULL)
           d += b;
       if (d >= m) d -= m;
       a <<= 1;
   }
   return d;
}


uint64_t pow_mod(uint64_t a, uint64_t b, uint64_t m)
{
    uint64_t r = m==1?0:1;
    while (b > 0) {
        if(b & 1)
            r = mul_mod(r, a, m);
        b = b >> 1;
        a = mul_mod(a, a, m);
    }
    return r;
}


int main()
{
    int a = 1777;
    for (int i = 0; i < 1855; ++i) {
        a = pow_mod(1777, a, 100000000);
    }
    printf("%d", a);
    return 0;
}

File: 188.py
def mul_mod(a: int, b: int, m: int) -> int:
    d = 0
    a %= m
    b %= m

    for i in range(64):
        d *= 2
        d %= m

        if a & 0x8000000000000000:
            d += b
        d %= m
        a *= 2

    return d


def pow_mod(a: int, b: int, m: int) -> int:
    r = m == 1 and 0 or 1
    while b > 0:
        if b & 1:
            r = mul_mod(r, a, m)
        b = b // 2
        a = mul_mod(a, a, m)
    return r


# print(pow_mod(2, 4, 5))
a = 1777
for i in range(1855):
    # a = pow_mod(1777, a, 100000000)
    a = pow(1777, a, 100000000)
print(a)
